### ISSUE TITLE
Sometimes, some tiles are not rendered

### ISSUE THREAD

**Issue by peterLaurence** [moagrius/TileView#252]
Hi,

Most of the time, tiles are correctly rendered. But sometimes, after an intense unzoom, some tiles are not rendered. At the beginning, i thought that my BitmapProvider was the culprit. Its implementation is very close to BitmapProviderAssets, because it looks for tiles located on the sd card. So i tried to log any Exception raised by BitmapFactory.decodeFile. But there were none.
At the moment, the only clue i have is that it happens just after a drop in memory usage of my app. I made a screen shot of that memory usage: 

![screen](https://cloud.githubusercontent.com/assets/15638794/12530271/c433df8c-c1d9-11e5-81ac-7df99e88002b.png)

As a side note, even if some tiles are not rendered, if i want them appear, i just have to scroll (or zoom / unzoom) a little to trigger tiles painting.

**Comment by moagrius** [comment]
Does this ever happen in the demo to you, or just in your custom implementation?

You can try calling `invalidate` or `postInvalidate` on the `TileView`, or on the `TileCanvasViewGroup` returned by `TileVIew.getTileCanvasViewGroup`

**Comment by peterLaurence** [comment]
Yes, it happens in the demo. Although it is not easy to trigger.
I chose the Large Image example (Joconde). What i do is :
• I zoom in the map to the maximum
• I unzoom out the map and i try to do it in one gesture (an "intense" unzoom). I don't make any gesture right after that.

After that, sometimes tiles are drawn correctly. But sometimes some of them are missing : 

![screenshot_20160124-113621](https://cloud.githubusercontent.com/assets/15638794/12535792/0f6579d8-c28f-11e5-89f5-ff856e3cc4a4.png)

To get back the missing tiles, we have to zoom in.

**Comment by moagrius** [comment]
Hm, have you tried `invalidate/postInvalidate`, per https://github.com/moagrius/TileView/issues/252#issuecomment-174218710?

**Comment by peterLaurence** [comment]
I don't understand why i should manually call invalidate/postInvalidate on the TileView object. It is not the upper layer's responsibility to tell the TileView to invalidate itself. It should do it automatically when it is necessary, which i think it does (i am not sure it is an invalidate issue).
I noticed that issue is most likely to happen with bigger maps than the ones in the demo. Right now, i am rendering maps with 6 DetailLevels, and i do severe tests with big zoom/unzoom, paning, ... 
The feeling i have right now is that the performance is sometimes excellent, and sometimes there are unexpected drops, like if the TileView was missing a little "cleanup". Because the performance at a given zoom level depends on what you did before :

• did you just came at zoom straight after loading the TileView ? => excellent performance
• did you zoom/unzoom like crazy, paned, etc.. => maybe performance will be affected, or maybe not
• if its a bit "laggy", calm down, let TileView breathe a little...  => performance is back again. 

I know it it not easy to investigate. But i am convinced that there is something behind this.

**Comment by moagrius** [comment]
I'm not suggesting you call `invalidate` as a permanent workaround, I'm trying to determine the cause and the solution.  If calling `invalidate` corrects the issue, then it's a pretty quick fix.  If it does not help, then I need to look elsewhere.

Does calling `invalidate/postInvalidate` correct the issue?

**Comment by peterLaurence** [comment]
Ok i will tell you that. Right now i can't do it (i have to go to bed, as i get up early to go to work :)
But tomorrow i test it and i tell you the result ;)

**Comment by moagrius** [comment]
:thumbsup:

**Comment by peterLaurence** [comment]
Calling invalidate don't help :(
To have better chance to see the issue (i insist that it rarely happens) :
• zoom to the maximum scale
• pan and don't wait the new tiles to be rendered : double-tap to go straight to the minimum scale

Repeat this until you see that not all tiles are rendered while you are on the minimum scale.

Btw and maybe not related to this issue, i saw that bnsantos talked in #250 about implementing parallelized tile loading. I don't know if he started to do something (or did you?). I have ideas on this and i could propose an implementation for this.

**Comment by moagrius** [comment]
It sounds like a drawing issue, since you said that dragging after causes them to render.  I have not seen this issue but will try to recreate as you've described.  i'll post back when i find something, and will leave the issue open until then.

regarding threads-per-tile, @bnsantos already implemented this in version 1, it just got lost in version 2 when i misunderstood how `AsyncTask.executeOnExecutor` worked.  he's recreating now.

thanks

**Comment by peterLaurence** [comment]
Ok. I have to be more precise when i told you that dragging after causes them to render. It seems to work if you drag enough : the missing tile have to get out of the screen. Although i am not 100% sure of this.

**Comment by moagrius** [comment]
cool, thanks for the info

**Comment by moagrius** [comment]
So despite repeated, very aggressive tries, I can't recreate the behavior on my Nexus 6.  What device are you using?  All my test devices are on loan - are you able to see this behavior in an emulator?  If so, which (built-in or genymotion), and what device images?

**Comment by peterLaurence** [comment]
I am using a Nexus 5, 2013 edition. And i have seen this only on this device. I just tried on the emulator (AVD, Nexus 5 api 23, Android studio 1.5.1) and i could not reproduce it. I also tried again on my real device, and i could reproduce it.
Well, it seems that i am not lucky :D

**Comment by moagrius** [comment]
I appreciate you posting back and taking the time to look into it.  Since I can't reproduce at the moment, I'm going to close the issue, but will definitely keep an eye out for other similar reports.

Thanks again

**Comment by vkourtis** [comment]
Hi, 

I have the same problem, with the difference that I am only using one bit tile.

I know what I'm doing is probably not the purpose of this library, but I am doing this as a temporary solution. Our system was designed to use one image that we would draw on, but we have started using very large images that I have to downsample in order to use them in the android app. So, I decided to start using tileview to present these single images and later move on to tiled images, when the rest of the system is ready to move on. The library is very slick and even using it for one image, made my code much clearer than my custom solution.

So what I have now is a tileview with only one tile, set up for example like this:
`setSize(1280, 800);`
`addDetailLevel(1f, imagePath, 1280, 800);`

Despite that difference, I have the same issue as the one described here. When I forcefully zoom out, the whole image disappears which, having only one tile, means that this one tile is not rendered.

I tried this on a Motorola Moto G (1st generation) and a Huawei P8 lite.

Any ideas? Is this really a bug, or caused by my possible "misuse" of the library?

**Comment by moagrius** [comment]
it's probably an out of memory issue - update your `BitmapProvider` to log out the exception that's being caught.  if it's OOME, then there's not much you can do - the device only has so much memory - in fact that's the whole point of `TileView`, to tile images and more aggressively manage memory consumption...

if you still need help, post back with the log information i asked for (and/or any other log information that's relevant)

**Comment by vkourtis** [comment]
Unfortunately, I get no exceptions. I'm catching all the possible exceptions in Bitmap Provider, but nothing reports back.

It all works fine (or almost fine) when I disable the memory cache on my Glide call. I can't find something robust to ask you on that, so I will just thank you for trying to help, even if not directly related to TileView :)

The problem I have now is that, at some zoom levels, I see my image two times. Like it tries to render a new tile right at the bottom of the first one.

Anyway, I know what I am trying to do is a bit strange. I will try to make it work somehow and later on convince my team to switch to tiled images...

**Comment by moagrius** [comment]
i mean i don't know for sure but if it's vanishing intermittently that makes me think it's probably a memory issue.

if you want to post your project - or a demo that shows the issue - i can try to take a look when i get some time, maybe in the next week or so

**Comment by vkourtis** [comment]
Hi again! I had this task on hold for some time, so now that I need to fix it, I'm coming back with more details. I'd be glad if you had the time to help me!

So I created a minimal project to showcase the problem and now I am actually seeing an exception, which is, in my case, a `java.lang.InterruptedException` coming from the `RequestFutureTarget` of the Glide library.

As you will see, what I am trying to do is create a custom view that is able to present a TileView and add some POIs on top of it with which the user can interact. But, as previously said, I am exploiting the TileView library to only use one image, for the time being.

You can find my project here https://github.com/vkourtis/ImageMapViewTest
To better explain my problem, I created this GIF (included in the project) https://github.com/vkourtis/ImageMapViewTest/blob/master/_showcase/tileview-double_image.gif

Thanks in advance!

**Comment by moagrius** [comment]
i'll take a look early next week

**Comment by vkourtis** [comment]
Thanks so much! Waiting for your advice!

**Comment by vkourtis** [comment]
After some months of abandoning this (no one was really using the feature), I tried this again. So, I managed to solve the problem with a tiny tiny hack :)

As I said before, I use a single tile (meaning the whole image). This means that the tile size is the same as the canvas size. So, I though that maybe this is an edge situation where the engine tries to create another tile because it thinks that it didn't fill out the whole canvas.

So, what I did is that I added 1 pixel in the tile size and... magic happened. Now when I zoom in and out, the second tile never appears. Also, there were still situations where the main tile would disappear and these are fixed as well.

To summarise, the change that fixed this was changing this line:
`this.addDetailLevel(1f, mImage, mImage.getTileWidth(), mImage.getTileHeight());`
to this:
`this.addDetailLevel(1f, mImage, mImage.getTileWidth() + 1, mImage.getTileHeight() + 1);`

I updated the code in the repo mentioned before with these changes. You will also find out that I tried updating the library, but this didn't change anything.

Hope this helps to figure out this edge situation, if this is something that you're willing to tackle.

**Comment by moagrius** [comment]
sorry for letting this languish - and thanks for posting the hint.  at the moment, all my energies are focused on v3, which is a top-to-bottom rewrite, so hopefully this issue gets fixed magically in the upgrade.  if not, it's still great to have this available for future searchers
